### PR TITLE
date: Add types

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds/fixes existing types for the `date` package.

There are some minor runtime changes but nothing major enough that I thought we should split this into several PRs. Hopefully this isn't too bad to review, happy to walk anyone through my changes though if y'all want it.

Part of #18838 

## How has this been tested?
Unit tests pass and type check passes.

## Types of changes
Non-breaking changes, just adding types 🙂 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by tyxla — COMMENTED** [review]
Nice work 👍  I've left some questions and nits but this looks great overall!

File: packages/date/src/index.js
Comment (sarayourfriend) [reply packages/date/src/index.js:41]: I only included the ones that are actually used in the `settings` object and in the `setSettings` function. Were the specific other ones you think we should include for some reason even though they're not used?

Comment (tyxla) [reply packages/date/src/index.js:41]: I think that's a good start 👍  The only problem would be if someone uses some of the other ones and gets TS errors. Those could be added incrementally, though - this isn't something that should be blocking this PR.

Comment (sarayourfriend) [reply packages/date/src/index.js:41]: Oh hmm that's interesting. I guess someone could be adding extraneous properties? But then shouldn't they indeed get an error considering those properties aren't supported by the code? Or should we extend this as a `Record<string, any>`?

Comment (tyxla) [reply packages/date/src/index.js:41]: Well, the thing is that they're supported by Moment, so practically they're also supported by the code. The typed ones we're adding here are only the default settings, right?

Anyway, we could make this a `Record`, but I believe would be much more precise if we just borrow the rest of the `LocaleSpecification` interface. 

I'll leave it to you to decide the level of precision you'd like to achieve with this first PR - I'm fine with what we have as a start already 👍

Comment (sarayourfriend) [reply packages/date/src/index.js:41]: > The typed ones we're adding here are only the default settings, right

Nope, if you check out the `setSettings` function I only added the properties we actually access off the `l10n` object.

Comment (tyxla) [reply packages/date/src/index.js:41]: Gotcha 👍 

Well, I don't see a good reason to add more in that case.

**Comment by sarayourfriend** [comment]
@tyxla Could this one get one last review from you?

**Review by tyxla — APPROVED** [review]
Nothing more to add. Looks great 🚢

**Pull request merged**
